### PR TITLE
Changes after the Mercury project

### PR DESCRIPTION
**Pull request by Dizzzmas**
## What does this PR do?
Brings a bunch of stuff we've added while building [Mercury](https://github.com/jetbridge/mercury) into the template repo.

TBD

### CODE REVIEW

**Comment by revmischa** [comment]
How do we rename it and remove the references to mercury?

**Comment by Dizzzmas** [comment]
> How do we rename it and remove the references to mercury?

I believe in sls-flask we've used a bash script for it. May try out the same thing here.

First I'd like to make sure that everything works locally as well as when deployed

**Comment by revmischa** [comment]
One fancy system I'm aware of is Yeoman https://yeoman.io/authoring/file-system.html
Not sure if there's anything newer/better
Or doing the same approach as in sls-flask is fine too. Was just a quick hack I put together

**Comment by Dizzzmas** [comment]
> One fancy system I'm aware of is Yeoman [yeoman.io/authoring/file-system.html](https://yeoman.io/authoring/file-system.html)
> Not sure if there's anything newer/better
> Or doing the same approach as in sls-flask is fine too. Was just a quick hack I put together

I'm always eager to try new stuff, but I think in this situation a better approach would be getting it to work with the bash script first and then improving on top of it. Someday we may try to replace it with Yeoman.


The draft of the script looks ready and works(at least for me):
```
git clone --depth=1 --branch=develop git@github.com:jetbridge/jkv2-ts-template.git myapp
npm run template:init
```

Next up is bringing the necessary changes to frontend and then trying to make it all work together. ReadMe also needs some work

File: README.md [hunk 1]
```diff
@@ -8,15 +8,15 @@ This includes JBCRA frontend, serverless backend with TypeORM connected to auror
...
-- `npm i -g npm@next-7`
+- `npm i -g npm@7`
 
 ## Quickstart
 
 ```shell
-npm i
-npm start
+git clone --depth=1 --branch=develop git@github.com:jetbridge/jkv2-ts-template.git myapp
+npm run template:init
```
Comment (revmischa) [inline README.md]: We should try this button https://github.com/jetbridge/jkv2-ts-template
Screen Shot 2020-12-08 at 12 17 23 AM

File: packages/backend/cloudformation/aurora-serverless.yml [hunk 2]
```diff
@@ -0,0 +1,114 @@
+Conditions:
```
Comment (revmischa) [inline packages/backend/cloudformation/aurora-serverless.yml]: I really hope we can hook up CDK someday

File: packages/backend/package.json [hunk 3]
```diff
@@ -23,6 +21,7 @@
     "aws-sdk": "^2.747.0",
     "aws-xray-sdk": "^3.2.0",
     "aws-xray-sdk-postgres": "^3.2.0",
+    "camelcase-keys": "^6.2.2",
```
Comment (revmischa) [inline packages/backend/package.json]: I think this is unused

File: packages/backend/package.json [hunk 4]
```diff
@@ -32,8 +31,10 @@
     "typeorm-aurora-data-api-driver": "^1.4.0"
   },
   "devDependencies": {
+    "@babel/helper-validator-option": "^7.12.0",
```
Comment (revmischa) [inline packages/backend/package.json]: what's this for?

File: packages/backend/src/api/listUsers.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { db } from ".."
+import { User, PaginatedResponse } from "jkv2-core"
+import { APIGatewayProxyEvent, APIGatewayProxyEventV2, APIGatewayProxyResultV2 } from "aws-lambda"
+import { getPaginationData, getPagesData } from "../util/pagination"
+
+/**
+ * List users
+ */
+export async function listUsers(event: APIGatewayProxyEventV2): Promise<APIGatewayProxyResultV2<PaginatedResponse<User>>> {
+  console.log(`LOCAL_DB: ${process.env.USE_LOCAL_DB}`)
```
Comment (revmischa) [inline packages/backend/src/api/listUsers.ts]:
```suggestion
```

File: packages/backend/src/api/listUsers.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  const pagesData = getPagesData(event.queryStringParameters)
```
Comment (revmischa) [inline packages/backend/src/api/listUsers.ts]: `getPaginated()` would be a better name

File: packages/backend/src/util/pagination.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+interface PaginatedResponse {
+    items?: any[]
+    paginationData?: PaginationData
+    statusCode?: number
+    message?: string
+}
+
+interface PagesData {
```
Comment (revmischa) [inline packages/backend/src/util/pagination.ts]:
```suggestion
interface PagedData {
```

File: packages/core/.env.local [hunk 8]
```diff
@@ -0,0 +1,3 @@
+USE_LOCAL_DB = true
+DB_NAME = jkv2
```
Comment (revmischa) [inline packages/core/.env.local:2]: this shouldn't be in git - should be ignored

File: packages/core/ormconfig.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { SnakeNamingStrategy } from "typeorm-naming-strategies/snake-naming.strategy"
+import { getConnectionOptions } from "./src/db/Connection"
```
Comment (revmischa) [inline packages/core/ormconfig.ts]: ormConfig.ts

File: packages/core/src/db/Connection.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+// list of entities from core go here
+const ALL_ENTITIES = [User]
+
+const CONNECTION_NAME = "default"
+
+/**
+ * Database manager class
+ * Taken from https://medium.com/safara-engineering/wiring-up-typeorm-with-serverless-5cc29a18824f
+ * This MAY NOT BE THE BEST WAY TO GET/CACHE DB CONNECTIONS!
```
Comment (revmischa) [inline packages/core/src/db/Connection.ts]: this code may be totally wrong

File: packages/core/src/db/Connection.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        return connection
+    }
+}
+
+export function getConnectionOptions(): ConnectionOptions {
+    let connectionOptions: ConnectionOptions
+    let logging: LoggerOptions
+    if (process.env.SQL_ECHO) {
```
Comment (revmischa) [inline packages/core/src/db/Connection.ts]: should be documented

File: packages/core/src/db/Connection.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            throw new Error("Couldn't get RDS ARNs from environment.")
+        }
+        // aurora sls
+        connectionOptions = {
+            entities: ALL_ENTITIES,
+            type: "aurora-data-api-pg",
+            database: "jkv2",
+            secretArn: process.env.RDS_SECRET_ARN,
+            resourceArn: process.env.RDS_ARN,
+            region: "us-east-1",
```
Comment (revmischa) [inline packages/core/src/db/Connection.ts]: don't hardcode this

File: packages/core/src/db/Connection.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            logging: logging, // log queries
+            name: CONNECTION_NAME,
+            namingStrategy: new SnakeNamingStrategy(),
+        }
+    }
+
+    // Don't need a pwd locally
```
Comment (revmischa) [inline packages/core/src/db/Connection.ts]: could need a password locally

File: packages/core/src/util/pagination.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
+
+interface PaginationData {
```
Comment (revmischa) [inline packages/core/src/util/pagination.ts]: looks like the interfaces from the other file

File: packages/frontend/package.json [hunk 15]
```diff
@@ -3,34 +3,33 @@
...
+    "@material-ui/core": "^4.11.0",
     "@material-ui/icons": "latest",
     "@material-ui/styles": "latest",
-    "@testing-library/dom": "^7.24.3",
-    "@testing-library/jest-dom": "^5.11.4",
-    "@testing-library/react": "^9.5.0",
-    "@testing-library/user-event": "^7.2.1",
-    "@types/jest": "^24.9.1",
-    "@types/react-dom": "^16.9.8",
+    "aws-amplify": "^3.3.3",
```
Comment (revmischa) [inline packages/frontend/package.json]: we have amplify and cognito deps, should cognito be included in here? maybe?

File: packages/core/ormconfig.ts (hunk 9 above)
Comment (Dizzzmas) [reply packages/core/ormconfig.ts]: I think it's a convention in TypeORM to name this file in lower-case only

https://github.com/typeorm/typeorm/blob/master/docs/using-ormconfig.md

File: packages/core/.env.local (hunk 8 above)
Comment (Dizzzmas) [reply packages/core/.env.local:2]: Can't really understand why. 

This file provides the required variables for developing locally right away. Yes, for some environments what I've written there may not work(For example port numbers etc.) I think that's still better than making people create `.env.local` and then consult the README or guess what vars to put there

Comment (revmischa) [reply packages/core/.env.local:2]: Then why not put those default settings in .env or .env.development?
Please read this doc https://create-react-app.dev/docs/adding-custom-environment-variables#adding-development-environment-variables-in-env

**Pull request closed without merging**